File: stdlib/lua/hull/tests/test_form.lua
-- test_form.lua — Tests for hull.form
--
-- Tests pure-function URL-encoded body parsing (no runtime globals needed).
-- Run via: the C test harness (test_lua_runtime.c) loads and executes this.

local form = require('hull.form')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- ── basic parsing ────────────────────────────────────────────────────

test("simple key-value pair", function()
    local r = form.parse("name=alice")
    assert_eq(r.name, "alice")
end)

test("multiple pairs", function()
    local r = form.parse("email=a%40b.com&pass=secret")
    assert_eq(r.email, "a@b.com")
    assert_eq(r.pass, "secret")
end)

test("plus decoded to space", function()
    local r = form.parse("msg=hello+world")
    assert_eq(r.msg, "hello world")
end)

test("percent decoding", function()
    local r = form.parse("q=%E4%BD%A0%E5%A5%BD")
    assert_eq(r.q, "\xE4\xBD\xA0\xE5\xA5\xBD")
end)

-- ── edge cases ───────────────────────────────────────────────────────

test("nil returns empty table", function()
    local r = form.parse(nil)
    assert(next(r) == nil, "expected empty table")
end)

test("empty string returns empty table", function()
    local r = form.parse("")
    assert(next(r) == nil, "expected empty table")
end)

test("non-string returns empty table", function()
    local r = form.parse(42)
    assert(next(r) == nil, "expected empty table")
end)

test("pair without = is skipped", function()
    local r = form.parse("noequals&key=val")
    assert_eq(r.key, "val")
    assert_eq(r.noequals, nil)
end)

test("empty key is skipped", function()
    local r = form.parse("=value&key=val")
    assert_eq(r[""], nil)
    assert_eq(r.key, "val")
end)

test("empty value is preserved", function()
    local r = form.parse("key=")
    assert_eq(r.key, "")
end)

test("duplicate keys: last wins", function()
    local r = form.parse("a=1&a=2&a=3")
    assert_eq(r.a, "3")
end)

test("value with = sign", function()
    local r = form.parse("data=a=b=c")
    assert_eq(r.data, "a=b=c")
end)

test("multiple & separators", function()
    local r = form.parse("a=1&&b=2")
    assert_eq(r.a, "1")
    assert_eq(r.b, "2")
end)

test("key with percent encoding", function()
    local r = form.parse("my%20key=value")
    assert_eq(r["my key"], "value")
end)

-- Return results for C test harness
return {pass = pass, fail = fail}

File: stdlib/js/hull/tests/test_form.js
// test_form.js — Tests for hull:form
//
// Tests pure-function URL-encoded body parsing (no runtime globals needed).

import { form } from "hull:form";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + e.message);
    }
}

function assertEq(a, b, msg) {
    if (a !== b)
        throw new Error((msg || "") + " expected " + b + ", got " + a);
}

// ── basic parsing ────────────────────────────────────────────────────

test("simple key-value pair", () => {
    const r = form.parse("name=alice");
    assertEq(r.name, "alice");
});

test("multiple pairs", () => {
    const r = form.parse("email=a%40b.com&pass=secret");
    assertEq(r.email, "a@b.com");
    assertEq(r.pass, "secret");
});

test("plus decoded to space", () => {
    const r = form.parse("msg=hello+world");
    assertEq(r.msg, "hello world");
});

// ── edge cases ───────────────────────────────────────────────────────

test("null returns empty object", () => {
    const r = form.parse(null);
    assertEq(Object.keys(r).length, 0);
});

test("empty string returns empty object", () => {
    const r = form.parse("");
    assertEq(Object.keys(r).length, 0);
});

test("non-string returns empty object", () => {
    const r = form.parse(42);
    assertEq(Object.keys(r).length, 0);
});

test("pair without = is skipped", () => {
    const r = form.parse("noequals&key=val");
    assertEq(r.key, "val");
    assertEq(r.noequals, undefined);
});

test("empty key is skipped", () => {
    const r = form.parse("=value&key=val");
    assertEq(r[""], undefined);
    assertEq(r.key, "val");
});

test("empty value is preserved", () => {
    const r = form.parse("key=");
    assertEq(r.key, "");
});

test("duplicate keys: last wins", () => {
    const r = form.parse("a=1&a=2&a=3");
    assertEq(r.a, "3");
});

test("value with = sign", () => {
    const r = form.parse("data=a=b=c");
    assertEq(r.data, "a=b=c");
});

test("multiple & separators", () => {
    const r = form.parse("a=1&&b=2");
    assertEq(r.a, "1");
    assertEq(r.b, "2");
});

test("key with percent encoding", () => {
    const r = form.parse("my%20key=value");
    assertEq(r["my key"], "value");
});

export default { pass, fail };
